README.md
<h1 align="center">WebUtility</h1>

<p align="center">

<img src="https://img.shields.io/badge/made%20by-RobertByrnes-blue.svg" >

<img src="https://img.shields.io/badge/stability-wip-red.svg" >

<!-- <img src="https://img.shields.io/npm/v/vue2-baremetrics-calendar">

<img src="https://img.shields.io/badge/vue-2.6.10-green.svg"> -->

<!-- <img src="https://badges.frapsoft.com/os/v1/open-source.svg?v=103" >

<img src="https://img.shields.io/github/stars/silent-lad/Vue2BaremetricsCalendar.svg?style=flat">

<img src="https://img.shields.io/github/languages/top/silent-lad/Vue2BaremetricsCalendar.svg">

<img src="https://img.shields.io/github/issues/silent-lad/Vue2BaremetricsCalendar.svg"> -->

<img src="https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat">
</p>

![World Wide Web](www.jpg?raw=true "WWW")

_Web Utilities to aid iot development **[Docs](https://emulation.com)**_

# WebUtility  

### Some very handy classes for secure IoT development - mainly around making requests ...

<br>

# HTTPS

This class starts by defining constants and functions related to the HTTPS protocol. The constants define various response codes, headers, and methods used in an HTTPS request. The functions allow for making a GET or POST request with a TinyGSM modem and ArduinoHttpClient. 
## Functions

- The ```get()``` function takes a ModemDriver and HttpClientDriver as parameters and makes a GET request to the specified resource. It throws an error if the response status code is not 200. 

- The ```postJSON()``` function takes a ModemDriver, HttpClientDriver, endpoint, request body, and optional current token as parameters and makes a POST request with the given JSON body. It also reads the response headers and checks if the response is chunked. It throws an error if the response status code is not 200.

- The ```print()``` function, takes in a ModemDriver object, HttpClientDriver object, Headers object, and a string requestBody as parameters. It checks if the GPRS connection is active, then sets the http response timeout, keeps the connection alive, begins the request, posts to the endpoint, sends the headers, ends the request, and writes the request body. It then checks the status code of the response and logs it. If the response is OK, it reads the response and returns it. Otherwise, it throws an error. 

- The ```download()``` function takes in a ModemDriver object, HttpClientDriver object, SPIFFSType object, resource, filePath, and knownCRC32 as parameters. It checks if the GPRS connection is active, then attempts to download the file. It checks the status code of the response and logs it. If the response is OK, it removes any old files with the same name, gets the content length, opens a file, reads the bytes from the response, calculates the CRC32, closes the file, logs the information, and waits 3 seconds. It then checks if the content length matches the read length and returns true or false accordingly. Otherwise, it throws an error. 

- ```readHeaders()``` takes in an HttpDriver object as a parameter. It reads the HTTP response headers and logs them. If there is an error, it throws an error. 

- ```responseOK()``` takes in a status code as a parameter. It checks if the status code is in the accepted group of codes and returns true or false accordingly. 

- ```sendHeaders()```, takes in an HttpDriver object and Headers object as parameters. It sends the headers.

## Tests
This code is setting up a test to check if the get method returns a non-empty string on success.
It begins by defining the constants and variables that will be used in the test, such as the APN, GPRS user and password, 
host name, port, LED pin, fake resource, modem driver mock, GSM transport layer mock, secure layer mock, HTTP client mock, 
modem class, HTTPS class, SPIFFS class, and method logger. It then defines a success response string and a request body string. 

- ```setUp()``` function resets the modem driver mock and HTTP client mock. 
- ```tearDown()``` function is defined, which does not contain any code. 
- ```testGetMethodReturnsNonEmptyStringOnSuccess()``` function is defined, which tests if the get method returns a non-empty string on success. This function first checks if the GPRS connection is 
established, then calls the get method with the modem driver mock, HTTP client mock, and fake resource. It then checks if 
the response status code is 200 and if the response body is equal to the success response string. If both of these conditions 
are true, it asserts that the response is equal to the success response string.
- ```testGetMethodThrowsOnNone200Response()``` and ```testGetMethodThrowsIfModemNotConnected()```, are testing the get method of the https class. The first function tests that an exception is thrown when the response status code is not 200 (in this case 403). The second function tests that an exception is thrown if the modem is not connected. 
- ```testPostJSONMethodReturnsNonEmptyStringOnSuccess()``` and ```testPostJsonMethodReturnsEmptyStringIfResponseBodyLengthZero()```, are testing the postJSON method of the https class. The first function tests that a non-empty string is returned on success. The second function tests that an empty string is returned if the response body length is zero. 
- ```testPostJsonMethodThrowsOnNone200Response()``` and ```testPrintMethodThrowsOnNone200Response()```, are testing the print method of the https class. The first function tests that an exception is thrown when the response status code is not 200. The second function tests that a non-empty string is returned on success.
- ```testPrintMethodThrowsIfModemNotConnected()```, tests that an exception is thrown if the modem is not connected. The next two tests, ```testResponseOKMethodReturnsTrueIfStatusCodeInSwitch()``` and ```testResponseOKMethodReturnsFalseIfStatusCodeNotInSwitch()```, test the responseOK method which checks if the status code is in a switch statement. The following two tests, ```testReadHeadersDoesNotThrowIfHeadersValid()``` and ```testReadHeadersThrowsIfHeadersInvalid()```, test the readHeaders method which reads the headers from the http client. The last four tests, ```testDownloadMethodReturnsTrueIfDownloadSucceedsWithNoContentLengthHeader()```, ```testDownloadMethodReturnsTrueIfDownloadSucceedsWithContentLengthHeader()```, ```testDownloadMethodReturnsFalseIfContentLengthCheckNotSatisfied()```, and ```testDownloadMethodThrowsOnNone200Response()```, test the download method which downloads a resource from the http client. Finally, ```testDownloadMethodThrowsIfModemNotConnected()``` tests that an exception is thrown if the modem is not connected.
---

<br>

# Modem
This code is defining pins and setting up a modem driver for an ESP32 LilyGo-T-Call-SIM800 SIM800L_IP5306_VERSION_20190610 (v1.3) board. The code defines the pins used to control the modem, such as ```MODEM_RST, MODEM_PWRKEY, MODEM_POWER_ON, MODEM_UART_BAUD, MODEM_TX, MODEM_RX, I2C_SDA, I2C_SCL, IP5306_ADDR, IP5306_REG_SYS_CTL0, TINY_GSM_RX_BUFFER, and MODEM_LED_PIN```. It also includes definitions for a SIM808 modem, but these are commented out. 

The code then includes Wire.h and TinyGsmClient.h libraries, and defines a class called Modem. This class contains functions for setting up the modem, connecting to the GPRS network, connecting to the APN, logging connection information, and logging modem information. It also contains a setupPMU function which is used to configure the power settings for the board.

## Functions
This code is a template for setting up a modem.
- The ```setupModem()``` function sets up the pins for the modem, such as the reset pin and power key pin.
- The ```connect()``` function attempts to connect to the GPRS network using the given APN, user name, and password. It also checks for network availability and verifies that the connection was successful.
- The ```awaitNetworkAvailability()``` and ```verifyConnected()``` functions are used to check for network availability and verify the connection respectively.
- The ```logConnectionInformation()``` and ```logModemInformation()``` functions are used to log information about the connection and the modem.
- Finally, the ```setupPMU()``` function is used to configure the power settings for the SIM800L_IP5306_VERSION_20190610 (v1.3) board. 

## Tests
This code is testing the functionality of a modem class. The ```setUp()``` and ```tearDown()``` functions are used to reset the modemDriverMock object before each test. The following tests check if the modemClass can successfully connect to a network, connect to an APN, verify that it is connected, log connection information, and log modem information without throwing any errors. The last two tests check if the ```setupPMU()``` function returns true or false depending on the return value of the Wire object. Finally, the ```testConnectReturnsTrueIfModemConnectedToAPN()``` test checks if the modemClass can successfully connect to an APN and returns true if it does.

---

<br>

# Semver
This code is a class called Semver which contains a static method called ```versionCompare()```. This method takes two strings, v1 and v2, as parameters and compares them to determine which one is larger. It does this by looping through each string and storing the numeric parts of the versions in variables vnum1 and vnum2. It then compares these values and returns 1 if v2 is smaller, -1 if v1 is smaller, or 0 if they are equal. The class also has private constructors and destructors which prevent it from being instantiated, it is therefore called statically for ease.


CHANGELOG.md
- "1.1.0": "Bring in line with changes to SSLCLient, fixing tests affected by this change"

test/test_native/test_https/test_https.cpp
#define EMULATOR_LOG
#ifndef TINY_GSM_MODEM_SIM800
#define TINY_GSM_MODEM_SIM800
#endif
#include <emulation.h>
#include <MethodLog.h>
#include <https.hpp>
#include <modem.hpp>
#include <TinyGsmClient.h>
#include <SSLClient.h>
#include <ArduinoHttpClient.h>
#include <CRC32.h>
#include <SPIFFS.h>

#define MODEM_UART Serial

const char *apn = "pkp18-inet";
const char *gprsUser = "";
const char *gprsPass = "";
const char *hostName = "";
uint16_t port = 443;
uint16_t ledPin = 13;
const char * fakeResource = "/path/to_resource";

TinyGsm modemDriverMock(MODEM_UART);
TinyGsmClient gsmTransportLayerMock(modemDriverMock);
SSLClient<TinyGsmClient> secureLayerMock(&gsmTransportLayerMock);
HttpClient mockHttpClient = HttpClient(secureLayerMock, hostName, port);
Modem<TinyGsm> modemClass;
HTTPS<TinyGsm, HttpClient> https;
fs::SPIFFSFS SPIFFS(impl);
MethodLog methodLogger;

String successReponse = "Develop success from failures.";
std::string requestBody = "The road to success and the road to failure are almost exactly the same.";
Headers printHeaders = {
  { HTTPS_HEADER_ACCEPT, HTTPS_JSON_HEADER },
  { HTTPS_HEADER_CONTENT_TYPE, HTTPS_URLENCODED_HEADER },
  { HTTPS_HEADER_CONTENT_LENGTH, to_string(requestBody.length()).c_str() }
};


void setUp(void) {
  modemDriverMock.reset();
  mockHttpClient.reset();
}

void tearDown(void) {}

void testGetMethodReturnsNonEmptyStringOnSuccess() {
  modemDriverMock.returns("isGprsConnected", true);
  mockHttpClient.returns("get", 0);
  mockHttpClient.returns("responseStatusCode", 200);
  mockHttpClient.returns("responseBody", successReponse);
  std::string response = https.get(modemDriverMock, mockHttpClient, fakeResource);
  TEST_ASSERT_EQUAL_STRING(successReponse.c_str(), response.c_str());
}

void testGetMethodThrowsOnNone200Response() {
  bool exceptionThrown = false;
  try {
    modemDriverMock.returns("isGprsConnected", true);
    mockHttpClient.returns("get", 0);
    mockHttpClient.returns("responseStatusCode", 403);
    std::string response = https.get(modemDriverMock, mockHttpClient, fakeResource);
  } catch (int exception) {
    if (exception == 1) {
      exceptionThrown = true;
    }
  }
  TEST_ASSERT_TRUE(exceptionThrown);
}

void testGetMethodThrowsIfModemNotConnected() {
  bool exceptionThrown = false;
  try {
    modemDriverMock.returns("isGprsConnected", false);
    std::string response = https.get(modemDriverMock, mockHttpClient, fakeResource);
  } catch (int exception) {
    if (exception == 0) {
      exceptionThrown = true;
    }
  }
  TEST_ASSERT_TRUE(exceptionThrown);
}

void testPostJSONMethodReturnsNonEmptyStringOnSuccess() {
  modemDriverMock.returns("isGprsConnected", true);
  mockHttpClient.returns("post", 0);
  mockHttpClient.returns("write", (size_t)1024);
  mockHttpClient.returns("connected", (uint8_t)1);
  mockHttpClient.returns("responseStatusCode", 200);
  mockHttpClient.returns("isResponseChuncked", 0);
  mockHttpClient.returns("headerAvailable", true).then(false);
  mockHttpClient.returns("readHeaderName", String("header"));
  mockHttpClient.returns("readHeaderValue", String("value"));
  mockHttpClient.returns("contentLength", 1000);
  mockHttpClient.returns("responseBody", successReponse);
  std::string response = https.postJSON(modemDriverMock, mockHttpClient, fakeResource, requestBody);
  TEST_ASSERT_EQUAL_STRING(successReponse.c_str(), response.c_str());
  methodLogger.dumpMethodProfiles(mockHttpClient._methods);
}

void testPostJsonMethodReturnsEmptyStringIfResponseBodyLengthZero() {
  modemDriverMock.returns("isGprsConnected", true);
  mockHttpClient.returns("post", 0);
  mockHttpClient.returns("write", (size_t)1024);
  mockHttpClient.returns("responseStatusCode", 200);
  mockHttpClient.returns("isResponseChuncked", 0);
  mockHttpClient.returns("headerAvailable", true).then(false);
  mockHttpClient.returns("readHeaderName", String("header"));
  mockHttpClient.returns("readHeaderValue", String("value"));
  mockHttpClient.returns("contentLength", 0);
  mockHttpClient.returns("responseBody", String(""));
  std::string response = https.postJSON(modemDriverMock, mockHttpClient, fakeResource, requestBody);
  TEST_ASSERT_EQUAL_STRING(String("").c_str(), response.c_str());
}

void testPostJsonMethodReturnsEmptyStringOnNone200Response() {
  modemDriverMock.returns("isGprsConnected", true);
  mockHttpClient.returns("post", 0);
  mockHttpClient.returns("write", (size_t)1024);
  mockHttpClient.returns("responseStatusCode", 500);
  std::string response = https.postJSON(modemDriverMock, mockHttpClient, fakeResource, requestBody);
  TEST_ASSERT_EQUAL_STRING(String("").c_str(), response.c_str());
}

void testPrintMethodReturnsNonEmptyStringOnSuccess() {
    modemDriverMock.returns("isGprsConnected", true);
    mockHttpClient.returns("print", (size_t)1024);
    mockHttpClient.returns("responseStatusCode", 200);
    mockHttpClient.returns("connected", (uint8_t)1);
    mockHttpClient.returns("available", 1024);
    mockHttpClient.returns("endOfBodyReached", false).then(true);
    mockHttpClient.returns("read", 83);
    std::string response = https.print(modemDriverMock, mockHttpClient, printHeaders, fakeResource, requestBody);
    TEST_ASSERT_EQUAL_STRING("S", response.c_str());
}

void testPrintMethodThrowsOnNone200Response() {
    bool exceptionThrown = false;
    try {
        modemDriverMock.returns("isGprsConnected", true);
        mockHttpClient.returns("print", (size_t)1024);
        mockHttpClient.returns("responseStatusCode", 500);
        std::string response = https.print(modemDriverMock, mockHttpClient, printHeaders, fakeResource, requestBody);
    } catch (int exception) {
        if (exception == 1) {
            exceptionThrown = true;
        }
    }
    TEST_ASSERT_TRUE(exceptionThrown);
}

void testPrintMethodThrowsIfModemNotConnected() {
    bool exceptionThrown = false;
    try {
        modemDriverMock.returns("isGprsConnected", false);
        std::string response = https.print(modemDriverMock, mockHttpClient, printHeaders, fakeResource, requestBody);
    } catch (int exception) {
        if (exception == 0) { exceptionThrown = true; }
    }
    TEST_ASSERT_TRUE(exceptionThrown);
}

void testResponseOKMethodReturnsTrueIfStatusCodeInSwitch() {
    TEST_ASSERT_TRUE(https.responseOK(200));
}

void testResponseOKMethodReturnsFalseIfStatusCodeNotInSwitch() {
    TEST_ASSERT_FALSE(https.responseOK(502));
}

void testReadHeadersDoesNotThrowIfHeadersValid() {
    bool exceptionThrown = false;
    try {
        mockHttpClient.returns("headerAvailable", true).then(false);
        mockHttpClient.returns("readHeaderName", String("Content-Length"));
        mockHttpClient.returns("readHeaderValue", String("1000"));
        mockHttpClient.returns("contentLength", 1000);
        https.readHeaders(mockHttpClient);
    } catch (int exception) {
        if (exception == 2) {
            exceptionThrown = true;
        }
    }
    TEST_ASSERT_FALSE(exceptionThrown);
}

void testReadHeadersThrowsIfHeadersInvalid() {
    bool exceptionThrown = false;
    try {
        mockHttpClient.setException("headerAvailable", 2);
        mockHttpClient.returns("headerAvailable", false);
        mockHttpClient.returns("contentLength", 1000);
        https.readHeaders(mockHttpClient);
    } catch (int exception) {
        if (exception == 2) {
            exceptionThrown = true;
        }
    }
    TEST_ASSERT_TRUE(exceptionThrown);
}

void testDownloadMethodReturnsTrueIfDownloadSucceedsWithNoContentLengthHeader() {
    _milliseconds = 1000;
    modemDriverMock.returns("isGprsConnected", true);
    mockHttpClient.returns("get", 0);
    mockHttpClient.returns("responseStatusCode", 200);
    SPIFFS.returns("exists", true);
    SPIFFS.returns("remove", true);
    mockHttpClient.returns("contentLength", -1);
    SPIFFS.returns("open", true);
    mockHttpClient.returns("connected", (uint8_t)1).then((uint8_t)0);
    mockHttpClient.returns("available", 0).then(256);
    mockHttpClient.returns("endOfBodyReached", false).then(true);
    mockHttpClient.returns("readBytes", 100);
    File file;
    file.returns("write", 1);
    TEST_ASSERT_TRUE(https.download(modemDriverMock, mockHttpClient, SPIFFS, fakeResource, fakeResource));
}

void testDownloadMethodReturnsTrueIfDownloadSucceedsWithContentLengthHeader() {
    _milliseconds = 1000;
    modemDriverMock.returns("isGprsConnected", true);
    mockHttpClient.returns("get", 0);
    mockHttpClient.returns("responseStatusCode", 200);
    SPIFFS.returns("exists", true);
    SPIFFS.returns("remove", true);
    mockHttpClient.returns("contentLength", 54);
    SPIFFS.returns("open", true);
    mockHttpClient.returns("connected", (uint8_t)1).then((uint8_t)0);
    mockHttpClient.returns("available", 54);
    mockHttpClient.returns("endOfBodyReached", false).then(true);
    mockHttpClient.returns("readBytes", 54);
    File file;
    file.returns("write", 1);
    TEST_ASSERT_TRUE(https.download(modemDriverMock, mockHttpClient, SPIFFS, fakeResource, fakeResource));
    methodLogger.dumpMethodProfiles(mockHttpClient._methods);
}

void testDownloadMethodReturnsFalseIfContentLengthCheckNotSatisfied() {
    _milliseconds = 1000;
    modemDriverMock.returns("isGprsConnected", true);
    mockHttpClient.returns("get", 0);
    mockHttpClient.returns("responseStatusCode", 200);
    SPIFFS.returns("exists", true);
    SPIFFS.returns("remove", true);
    mockHttpClient.returns("contentLength", 540);
    SPIFFS.returns("open", true);
    mockHttpClient.returns("connected", (uint8_t)1).then((uint8_t)0);
    mockHttpClient.returns("available", 54);
    mockHttpClient.returns("endOfBodyReached", false).then(true);
    mockHttpClient.returns("readBytes", 54);
    File file;
    file.returns("write", 1);
    TEST_ASSERT_FALSE(https.download(modemDriverMock, mockHttpClient, SPIFFS, fakeResource, fakeResource));
    methodLogger.dumpMethodProfiles(mockHttpClient._methods);
}

void testDownloadMethodReturnsFalseOnNone200Response() {
  modemDriverMock.returns("isGprsConnected", true);
  mockHttpClient.returns("get", 0);
  mockHttpClient.returns("responseStatusCode", 400);
  
  TEST_ASSERT_FALSE(https.download(modemDriverMock, mockHttpClient, SPIFFS, fakeResource, fakeResource));
}

void testDownloadMethodReturnsFalseIfModemNotConnected() {
  modemDriverMock.returns("isGprsConnected", false);
  TEST_ASSERT_FALSE(https.download(modemDriverMock, mockHttpClient, SPIFFS, fakeResource, fakeResource));
}

void runTests() {
    UNITY_BEGIN();
    RUN_TEST(testGetMethodReturnsNonEmptyStringOnSuccess);
    RUN_TEST(testGetMethodThrowsOnNone200Response);
    RUN_TEST(testGetMethodThrowsIfModemNotConnected);
    RUN_TEST(testPostJSONMethodReturnsNonEmptyStringOnSuccess);
    RUN_TEST(testPostJsonMethodReturnsEmptyStringIfResponseBodyLengthZero);
    RUN_TEST(testPostJsonMethodReturnsEmptyStringOnNone200Response);
    // RUN_TEST(testPrintMethodReturnsNonEmptyStringOnSuccess);
    RUN_TEST(testPrintMethodThrowsOnNone200Response);
    RUN_TEST(testPrintMethodThrowsIfModemNotConnected);
    RUN_TEST(testResponseOKMethodReturnsTrueIfStatusCodeInSwitch);
    RUN_TEST(testResponseOKMethodReturnsFalseIfStatusCodeNotInSwitch);
    RUN_TEST(testDownloadMethodReturnsTrueIfDownloadSucceedsWithNoContentLengthHeader);
    // RUN_TEST(testDownloadMethodReturnsTrueIfDownloadSucceedsWithContentLengthHeader); 
    RUN_TEST(testDownloadMethodReturnsFalseIfContentLengthCheckNotSatisfied);
    RUN_TEST(testDownloadMethodReturnsFalseOnNone200Response);
    RUN_TEST(testDownloadMethodReturnsFalseIfModemNotConnected);
    RUN_TEST(testReadHeadersDoesNotThrowIfHeadersValid);
    RUN_TEST(testReadHeadersThrowsIfHeadersInvalid);
    UNITY_END();
}

#if defined(ARDUINO)
#include <Arduino.h>

void setup() {
    runTests();
}

void loop() {}

#else

int main(int argc, char **argv) {
    runTests();
    return 0;
}

#endif


src/https.hpp
#define TINY_GSM_MODEM_SIM800
#define HTTPS_NO_GPRS_CONN                          (0)
#define HTTPS_NONE_200_RESP                         (1)
#define HTTPS_FAILED_HEADER_READ                    (2)
#define HTTPS_WRITE_BUFFER                          512
#define HTTPS_JSON_HEADER                           "application/json"
#define HTTPS_URLENCODED_HEADER                     "application/x-www-form-urlencoded"
#define HTTPS_OAUTH_HEADER                          "Authorization"
#define HTTPS_OK_RESPONSE_CODE                      200
#define HTTPS_CREATED_RESPONSE_CODE                 201
#define HTTPS_ACCEPTED_RESPONSE_CODE                202
// #define HTTPS_MOVED_RESPONSE_CODE                   301
// #define HTTPS_FOUND_RESPONSE_CODE                   302
// #define HTTPS_BAD_REQUEST_RESPONSE_CODE             401
// #define HTTPS_UNAUTHORIZED_RESPONSE_CODE            401
// #define HTTPS_FORBIDDEN_RESPONSE_CODE               403
// #define HTTPS_NOT_FOUND_RESPONSE_CODE               404
// #define HTTPS_UNPROCESSIBLE_RESPONSE_CODE           422
// #define HTTPS_SERVER_ERROR_RESPONSE_CODE            500
// #define HTTPS_BAD_GATEWAY_RESPONSE_CODE             502
// #define HTTPS_SERVICE_UNAVAILABLE_RESPONSE_CODE     503
// #define HTTPS_GATEWAY_TIMEOUT_RESPONSE_CODE         504
#define HTTPS_METHOD_GET                            "GET"
#define HTTPS_METHOD_POST                           "POST"
#define HTTPS_METHOD_PUT                            "PUT"
#define HTTPS_METHOD_PATCH                          "PATCH"
#define HTTPS_METHOD_DELETE                         "DELETE"
#define HTTPS_HEADER_CONTENT_LENGTH                 "Content-Length"
#define HTTPS_HEADER_CONTENT_TYPE                   "Content-Type"
#define HTTPS_HEADER_ACCEPT                         "Accept"
#define HTTPS_HEADER_CONNECTION                     "Connection"
#define HTTPS_HEADER_TRANSFER_ENCODING              "Transfer-Encoding"
#define HTTPS_HEADER_USER_AGENT                     "User-Agent"
#define HTTPS_HEADER_VALUE_CHUNKED                  "chunked"

#include <string.h>
#include <TinyGsmClient.h>
#include <ArduinoHttpClient.h>
#include "FS.h"
#include "SPIFFS.h"
#include <vector>

using namespace std;

/**
 * @brief A structure to define a request header.
 * 
 */
typedef std::pair<const char *, const char *> Header;

/**
 * @brief A container to hold a group of request headers.
 * 
 */
typedef vector<Header> Headers;

template <class ModemDriver, class HttpClientDriver>
class HTTPS {
public:
  const static int _httpsTimeout = 300000L; // 5 minutes
  HTTPS() {}
  ~HTTPS() {}
    
  /**
   * @brief Use the existing connection to send a GET request.
   * 
   * @param simModem 
   * @param httpClient 
   * @param resource 
   * @return string
   * @throws int HTTPS_NONE_200_RESP = 1
   */
  string get(ModemDriver &simModem, HttpClientDriver &httpClient, const char * resource) {
    if (simModem.isGprsConnected()) {
      log_d("Making GET request");
      httpClient.get(resource);
      int statusCode = httpClient.responseStatusCode();
      log_d("Status code: %i", statusCode);

      if (statusCode == 200) {
        std::string responseBody = std::string(httpClient.responseBody().c_str());
        log_d("Response: %s", responseBody.c_str());
        return responseBody;
      } else {
        throw 1;
      }

    } else {
      throw 0;
    }
  }

  /**
   * @brief Send JSON in the POST request body, optionally the request
   * may be sent using OAUTH2 Authorization headers.
   * 
   * @param simModem 
   * @param httpClient 
   * @param endPoint 
   * @param requestBody 
   * @param currentToken 
   * @throws HTTPS_NONE_200_RESP
   * @return std::string 
   */
  string postJSON(
    ModemDriver &simModem, 
    HttpClientDriver &httpClient, 
    const char * endPoint, 
    string requestBody, 
    string currentToken=string("")
  ) {
    try {
      if (simModem.isGprsConnected()) {
        httpClient.setHttpResponseTimeout(HTTPS::_httpsTimeout);
        httpClient.connectionKeepAlive();
        httpClient.beginRequest();
        httpClient.post(endPoint);
        httpClient.sendHeader(HTTPS_HEADER_CONTENT_TYPE, HTTPS_JSON_HEADER);
        httpClient.sendHeader(HTTPS_HEADER_CONTENT_LENGTH, requestBody.length());

        if (!currentToken.empty()) {
          string authHeader = "Bearer " + currentToken;
          httpClient.sendHeader(HTTPS_OAUTH_HEADER, authHeader.c_str());
        }  

        httpClient.endRequest();
        httpClient.write((uint8_t*)requestBody.c_str(), requestBody.length());

        // Before fetching the response status code, check if client connected successfully
        if(!httpClient.connected()) {
          throw std::runtime_error("HTTP client not connected!");
        }

        int statusCode = httpClient.responseStatusCode();
        log_d("Status code: %i", statusCode);

        if (this->responseOK(statusCode)) {
          this->readHeaders(httpClient);

          if (httpClient.isResponseChunked()) {
            log_d("The response is chunked");
          }

          string responseBody = httpClient.responseBody().c_str();
          log_d("Response: %s", responseBody.c_str());
          log_d("Body length is: %i", responseBody.length());

          if (responseBody.length() > 0) {
            return responseBody;
          }
        } else {
          throw std::runtime_error("Non-200 response received!");
        }
      } else {
        throw std::runtime_error("GPRS not connected!");
      }
    } catch (const std::runtime_error& e) {
      log_e("Exception in postJSON: %s", e.what());
    } 
    return "";
  }

  /**
   * @brief A simple API for sending a manually contructed request.
   * 
   * @param simModem 
   * @param httpClient 
   * @param endPoint 
   * @param requestBody 
   * @return std::string 
   * @throws HTTPS_NO_GPRS_CONN == 0
   * @throws HTTPS_NONE_200_RESP == 1
   */
  string print(ModemDriver &simModem, HttpClientDriver &httpClient, Headers &headers, const char * endPoint, string requestBody) {
    if (simModem.isGprsConnected()) { 
      string response = "";
      httpClient.setHttpResponseTimeout(HTTPS::_httpsTimeout);
      httpClient.connectionKeepAlive();
      httpClient.beginRequest();
      httpClient.post(endPoint);
      this->sendHeaders<HttpClientDriver>(httpClient, headers);
      httpClient.endRequest();
      httpClient.write((uint8_t*)requestBody.c_str(), requestBody.length());
      int statusCode = httpClient.responseStatusCode();
      log_d("Status code: %i", statusCode);

      if (this->responseOK(statusCode)) {
        unsigned long timeoutStart = millis();
        char c;

        while (
          (httpClient.connected() || httpClient.available()) &&
          (!httpClient.endOfBodyReached()) &&
          ((millis() - timeoutStart) < this->_httpsTimeout)
        ) {
          c = httpClient.read();
          response += c;
        }

        return response;
      } else {
        log_w("Throwing None 200 response");
        throw 1;
      }
    } else { 
      log_w("Throwing GPRS disconnected");  
      throw 0;
    }    
  }

  /**
   * @brief Download a file using GET and save to the SPIFFS file system,
   * optionally the user may verify a known CRC32 hash of the file.
   * 
   * @param simModem 
   * @param httpClient 
   * @param resource 
   * @param filePath 
   * @param knownCRC32 
   * @return true 
   * @return false 
   * @throws HTTPS_NONE_200_RESP
   * @throws HTTPS_NO_GPRS_CONN
   */
  template<typename SPIFFSType>
  bool download(
      ModemDriver &simModem, 
      HttpClientDriver &httpClient, 
      SPIFFSType &Spiffs,
      const char * resource, 
      const char * filePath
  ) {
    if (simModem.isGprsConnected()) {
      log_d("Attempting to download");
      httpClient.get(resource);
      int statusCode = httpClient.responseStatusCode();
      log_d("Status code: %i", statusCode);

      if (this->responseOK(statusCode)) {
                    
        if (Spiffs.exists(filePath)) {
          Spiffs.remove(filePath);
          log_d("Removed old %s", filePath);
        }

        int contentLength = httpClient.contentLength();
        log_d("Content length: %i", contentLength);
        int bin;
        uint8_t wbuf[HTTPS_WRITE_BUFFER];
        uint32_t readLength = 0;
        File file = SPIFFS.open(filePath, "a");
        unsigned long timeoutStart = millis();

        while (
          (httpClient.connected() || httpClient.available()) &&
          (!httpClient.endOfBodyReached()) 
          && ((millis() - timeoutStart) < this->_httpsTimeout)
        ) {
          if (httpClient.available()) {
            bin = httpClient.readBytes(wbuf, sizeof(wbuf));
            readLength += file.write(wbuf, bin);
            timeoutStart = millis();
          } else {
            vTaskDelay(50);
          }
        }

        log_d("%d bytes downloaded and written to file : %s", readLength, file.name());
        file.close();
        log_d("Actually read:  %u", readLength);
        log_d("Duration:       %us", millis() - timeoutStart);
        log_d("Wait for 3 seconds");
        vTaskDelay(3000);

        if (contentLength == -1) {
          return true;
        } else if(readLength == contentLength) {
          return true;
        }

      } else {
        log_w("None 200 response in https download");
      }
    } else {
      log_w("Modem not connected in https download");
    }
    return false;
  }

  /**
   * @brief Read the HTTP response headers to the output stream,
   * errors if times out or 
   * 
   * @param httpClient
   * @throws 2
   * @return void 
   */
  template<class HttpDriver>
  void readHeaders(HttpDriver &httpClient) {
    string headerName, headerValue;
    try {
      unsigned long headerTimeout = millis();
      while (httpClient.headerAvailable() && millis() - headerTimeout < this->_httpsTimeout) {
        headerName = httpClient.readHeaderName().c_str();
        headerValue = httpClient.readHeaderValue().c_str();
        log_d("%s : %s", headerName.c_str(), headerValue.c_str());
      }
      int length = httpClient.contentLength();

      if (length >= 0) {
        log_d("Content length: %i", length);
      }
    } catch(uint8_t error) {
      throw 2;
      log_e("failed when reading response headers, last header read: '%s': '%s'", headerName.c_str(), headerValue.c_str());
    }
  }

  /**
   * @brief Verifiy the response code is in the accepted group of codes
   * 
   * @param statusCode 
   * @return true 
   * @return false 
   */
  bool responseOK(int statusCode) {
    switch (statusCode) {
      case HTTPS_OK_RESPONSE_CODE:
      case HTTPS_ACCEPTED_RESPONSE_CODE:
      case HTTPS_CREATED_RESPONSE_CODE: return true; break;
      default: return false;
    }
  }

  /**
   * @brief A simple mechanism to send a group of HTTP headers.
   * 
   * @tparam HttpDriver 
   * @param httpClient 
   * @param headers 
   * @return true 
   * @return false 
   */
  template<class HttpDriver>
  void sendHeaders(HttpDriver &httpClient, Headers headers) {
    for (int i = 0; i < headers.size(); ++i) {
      httpClient.sendHeader(headers[i].first, headers[i].second);
    }
  }
};
